javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll

```diff
@@ -62,222 +95,465 @@ module ClientRequest {
- */
-private class FetchUrlRequest extends ClientRequest::Range {
-  DataFlow::Node url;
+    override DataFlow::Node getADataNode() {
+      exists(string name | name = "headers" or name = "body" | result = getOptionArgument(1, name))
+    }
 
-  FetchUrlRequest() {
-    exists(string moduleName, DataFlow::SourceNode callee | this = callee.getACall() |
-      (
-        moduleName = "node-fetch" or
-        moduleName = "cross-fetch" or
-        moduleName = "isomorphic-fetch"
-      ) and
-      callee = DataFlow::moduleImport(moduleName) and
-      url = getArgument(0)
-    )
-    or
-    this = DataFlow::globalVarRef("fetch").getACall() and
-    url = getArgument(0)
+    override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
+      responseType = "fetch.response" and
+      promise = true and
+      result = this
+    }
   }
 
-  override DataFlow::Node getUrl() { result = url }
+  /**
+  * A model of a URL request made using the `got` library.
+  */
+  class GotUrlRequest extends ClientRequest::Range {
+    GotUrlRequest() {
+      exists(string moduleName, DataFlow::SourceNode callee | this = callee.getACall() |
+        moduleName = "got" and
+        (
+          callee = DataFlow::moduleImport(moduleName) or
+          callee = DataFlow::moduleMember(moduleName, "stream")
+        )
+      )
+    }
+
+    override DataFlow::Node getUrl() {
+      result = getArgument(0) and
+      not exists(getOptionArgument(1, "baseUrl"))
+    }
+
+    override DataFlow::Node getHost() {
+      exists(string name |
+        name = "host" or
+        name = "hostname"
+      |
+        result = getOptionArgument(1, name)
+      )
+    }
 
-  override DataFlow::Node getHost() { none() }
+    override DataFlow::Node getADataNode() {
+      exists(string name | name = "headers" or name = "body" or name = "query" |
+        result = getOptionArgument(1, name)
+      )
+    }
 
-  override DataFlow::Node getADataNode() {
-    exists(string name | name = "headers" or name = "body" | result = getOptionArgument(1, name))
-  }
-}
+    /** Holds if the result is a stream. */
+    predicate isStream() {
+      getOptionArgument(1, "stream").mayHaveBooleanValue(true)
+      or
+      this = DataFlow::moduleMember("got", "stream").getACall()
+    }
 
-/**
- * A model of a URL request made using the `got` library.
- */
-private class GotUrlRequest extends ClientRequest::Range {
-  GotUrlRequest() {
-    exists(string moduleName, DataFlow::SourceNode callee | this = callee.getACall() |
-      moduleName = "got" and
+    /** Holds if the result is a JSON object. */
+    predicate isJson() {
+      getOptionArgument(1, "json").mayHaveBooleanValue(true)
+    }
+
+    override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
+      result = this and
       (
-        callee = DataFlow::moduleImport(moduleName) or
-        callee = DataFlow::moduleMember(moduleName, "stream")
+        isStream() and
```

Comment:
You could rephrase this as `if isStream() then ... else if isJson() ... else ...`, which would also model the fact that the `json` flag is [ignored](https://www.npmjs.com/package/got#json) in streaming mode.